plugins/verifier/cosign/cosign.go

```diff
@@ -78,84 +77,134 @@ func parseInput(stdin []byte) (*PluginInputConfig, error) {
 		}, nil
 	}
 
-	return &verifier.VerifierResult{
-		Name:      input.Config.Name,
-		IsSuccess: false,
-		Message:   "cosign verification failed. no valid signatures found",
-	}, nil
-}
-
-func signatures(ctx context.Context, img string, keyRef string, config *PluginInputConfig) (checkedSignatures []oci.Signature, bundleVerified bool, err error) {
-	registryClientOptions := []remote.Option{
-		remote.WithAuthFromKeychain(authn.DefaultKeychain),
-	}
-	var options []name.Option
-	if config.StoreWrapperConfig.StoreConfig.UseHttp {
-		options = append(options, name.Insecure)
-		// #nosec G402
-		registryClientOptions = append(registryClientOptions, remote.WithTransport(&http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}))
-	}
-
-	ref, err := name.ParseReference(img, options...)
+	subjectDesc, err := referrerStore.GetSubjectDescriptor(ctx, subjectReference)
 	if err != nil {
-		return nil, false, err
-	}
-
-	registryClientOptionsWrapper := []ociremote.Option{
-		ociremote.WithRemoteOptions(registryClientOptions...),
+		return &verifier.VerifierResult{
+			Name:      input.Config.Name,
+			IsSuccess: false,
+			Message:   fmt.Sprintf("cosign verification failed: failed to create subject hash: %v", err),
+		}, nil
 	}
-	cosignOpts := &cosign.CheckOpts{
-		ClaimVerifier:      cosign.SimpleClaimVerifier,
-		RegistryClientOpts: registryClientOptionsWrapper,
+	subjectDescHash := v1.Hash{
+		Algorithm: subjectDesc.Digest.Algorithm().String(),
+		Hex:       subjectDesc.Digest.Hex(),
 	}
 
-	var ecdsaVerifier signature.Verifier
-	var roots *x509.CertPool
-	if keyRef != "" {
-		ecdsaVerifier, err = loadPublicKey(ctx, keyRef)
+	signatures := []oci.Signature{}
+	for _, blob := range referenceManifest.Layers {
+		blobBytes, err := referrerStore.GetBlobContent(ctx, subjectReference, blob.Digest)
 		if err != nil {
-			return nil, false, err
+			return &verifier.VerifierResult{
+				Name:      input.Config.Name,
+				IsSuccess: false,
+				Message:   fmt.Sprintf("cosign verification failed: failed to get blob content: %v", err),
+			}, nil
 		}
-		cosignOpts.SigVerifier = ecdsaVerifier
-	} else {
-		roots, err = fulcio.GetRoots()
+		sig, err := static.NewSignature(blobBytes, blob.Annotations[static.SignatureAnnotationKey], staticLayerOpts(blob)...)
 		if err != nil {
-			return nil, false, err
+			return &verifier.VerifierResult{
+				Name:      input.Config.Name,
+				IsSuccess: false,
+				Message:   fmt.Sprintf("cosign verification failed: failed to generate static signature: %v", err),
+			}, nil
 		}
-		cosignOpts.RootCerts = roots
-		if cosignOpts.RootCerts == nil {
-			return nil, false, fmt.Errorf("failed to initialize root certificates")
+		// The verification will return an error if the signature is not valid.
+		// Currently the bundle verification result is ignored
+		_, err = cosign.VerifyImageSignature(ctx, sig, subjectDescHash, cosignOpts)
+		if err != nil {
```

Comment:
does the cosign lib also fails early when `VerifyImageSignature` failed?

Reply:
TLDR; yes it fails early. 

Cosign does 2 verifications: The first is the certificate verification (similar to notary) and the second is a 'bundle' verification which checks the authenticity of the signature. You can specify the Rekor URL to verify using the [transparency log](https://docs.sigstore.dev/security/#rekor-security-model). The verifier for the cert part returns an error which is propogated and back from VerifyImageSignature. The boolean bundleVerified return value is the result of the bundle verification (offline or online). We need to discuss how we are to incorporate the bundle verification value in the final verification result. For now, I've just ignored bundle verification and added a TODO.